File: java/src/main/java/global/thalion/ttio/protection/Verifier.java
/* TTI-O Java Implementation / Copyright (C) 2026 DTW-Thalion / SPDX-License-Identifier: LGPL-3.0-or-later */
package global.thalion.ttio.protection;

/**
 * High-level verification API.
 *
 * <p>Collapses the three outcomes of a sign-and-verify cycle
 * (valid / invalid / not-signed) into a single enum, plus an
 * {@link Status#ERROR} fallback for I/O failures. Use this instead
 * of {@link SignatureManager} directly when you want to render a
 * status to an end user.</p>
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOVerifier}, Python
 * {@code ttio.verifier.Verifier}.</p>
 *
 * @since 0.6
 */
public final class Verifier {

    /** Sign-and-verify cycle outcome. */
    public enum Status {
        VALID,
        INVALID,
        NOT_SIGNED,
        ERROR
    }

    private Verifier() {}

    /**
     * Verify a signature string against data and key, returning a
     * four-state status.
     *
     * @param data      original bytes (never {@code null})
     * @param signature signature string, or {@code null} / empty for {@code NOT_SIGNED}
     * @param key       32-byte HMAC key
     * @return status
     */
    public static Status verify(byte[] data, String signature, byte[] key) {
        if (signature == null || signature.isEmpty()) return Status.NOT_SIGNED;
        try {
            return SignatureManager.verify(data, signature, key)
                ? Status.VALID : Status.INVALID;
        } catch (RuntimeException e) {
            return Status.ERROR;
        }
    }
}

File: python/src/ttio/verifier.py
"""``Verifier`` — high-level sign/verify status wrapper."""
from __future__ import annotations

import base64
import hmac
from enum import IntEnum

from . import signatures


class VerificationStatus(IntEnum):
    """Sign-and-verify cycle outcome.

    Cross-language: ObjC ``TTIOVerificationStatus`` · Java
    ``global.thalion.ttio.protection.Verifier.Status``.
    """

    VALID = 0
    INVALID = 1
    NOT_SIGNED = 2
    ERROR = 3


class Verifier:
    """High-level verification API.

    Collapses the three outcomes of a sign-and-verify cycle
    (valid / invalid / not-signed) into a single enum, plus an error
    fallback for I/O failures. Use this instead of
    :mod:`ttio.signatures` when you want to render a status to an
    end user.

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIOVerifier`` · Java:
    ``global.thalion.ttio.protection.Verifier``.
    """

    @staticmethod
    def verify(data: bytes, signature: str | None, key: bytes) -> VerificationStatus:
        """Verify a signature string against data and key.

        Parameters
        ----------
        data : bytes
            Original bytes (never ``None``).
        signature : str or None
            Signature string from ``@ttio_signature`` or
            ``@provenance_signature``. ``None`` or empty returns
            :attr:`VerificationStatus.NOT_SIGNED`.
        key : bytes
            32-byte HMAC key.

        Returns
        -------
        VerificationStatus
        """
        if not signature:
            return VerificationStatus.NOT_SIGNED
        try:
            payload = (
                signature[len(signatures.SIGNATURE_V2_PREFIX):]
                if signature.startswith(signatures.SIGNATURE_V2_PREFIX)
                else signature
            )
            expected = signatures.hmac_sha256(data, key)
            actual = base64.b64decode(payload)
            if hmac.compare_digest(expected, actual):
                return VerificationStatus.VALID
            return VerificationStatus.INVALID
        except Exception:
            return VerificationStatus.ERROR
